move rpath to frontend and remove native paths from it

Jan200101 opened the pull request:
As discussed at compiler meeting [2023-10-26](https://github.com/ziglang/zig/wiki/Self-Hosted-Compiler-Meetings#2023-10-26).

Issues:
- recursive invocations of zig that pass arguments cause NativePaths to reappear in rpaths
  - `zig build` run `zig build-exe -L lib_dir` which adds `lib_dir` to the resulting rpath

Needed tests:
- [ ] system deps
- [ ] NixOS
- [ ] local deps

andrewrk commented:
This had conflicts with master branch, but it had already passed the CI before that, so I rebased it on your behalf.

I also dropped these commits from it:
 * https://github.com/ziglang/zig/pull/17917/commits/f306d2acdaefbf2db82c018e58a3dfbf49d321d2
 * https://github.com/ziglang/zig/pull/17917/commits/d4328e42f4c4eede8ef93125ea1a59890b71d2c6

I don't think the build script should be doing those things. Let's discuss those changes in a follow-up if you still want to pursue those.

Jan200101 commented:
> This had conflicts with master branch, but it had already passed the CI before that, so I rebased it on your behalf.

Thanks

> I also dropped these commits from it:
> 
>     * [f306d2a](https://github.com/ziglang/zig/commit/f306d2acdaefbf2db82c018e58a3dfbf49d321d2)
> 
>     * [d4328e4](https://github.com/ziglang/zig/commit/d4328e42f4c4eede8ef93125ea1a59890b71d2c6)
> 
> 
> I don't think the build script should be doing those things. Let's discuss those changes in a follow-up if you still want to pursue those.

These were mainly there to simplify testing the result of all the different environments without changing the overall behavior of Zig...

> One more thing: With just that NativePaths one-line change you made, it looks like it might be possible to change the default of each-lib-rpath to be off, or even have that option removed altogether.

It should be able to cover NixOS and systemwide libraries, but I am unsure if it would cover user libraries. Will need to do testing before that proposal can be made.
